tls/test_tls_handshake.py

```diff
@@ -465,3 +465,67 @@ def test(self):
         self.assertTrue(res, "Bad second handshake: %s" % res)
         res = x509_check_issuer(vhs.cert, "New Issuer")
         self.assertTrue(res, "Wrong certificate reloaded")
+
+class CloseConnection(tester.TempestaTest):
+
+    backends = [
+        {
+            'id' : '0',
+            'type' : 'deproxy',
+            'port' : '8000',
+            'response' : 'static',
+            'response_content' :
+                'HTTP/1.1 200 OK\r\n'
+                'Content-Length: 1\r\n\r\nx'
+        }
+    ]
+
+    tempesta = {
+        'config' : """
+            cache 0;
+            listen 443 proto=https;
+
+            tls_certificate ${general_workdir}/tempesta.crt;
+            tls_certificate_key ${general_workdir}/tempesta.key;
+
+            srv_group srv_grp1 {
+                server ${server_ip}:8000;
+            }
+            vhost tempesta-tech.com {
+                proxy_pass srv_grp1;
+            }
+            http_chain {
+                host == "tempesta-tech.com" -> tempesta-tech.com;
+                -> block;
+            }
+        """
+    }
+
+    def start_all(self):
+        self.sniffer = analyzer.AnalyzerCloseRegular(self.get_tempesta().node,
+                                                     self.get_tempesta().host,
+                                                     node_close=True,
+                                                     timeout=10, ports=(443,))
+        self.sniffer.start()
+
+        deproxy_srv = self.get_server('0')
+        deproxy_srv.start()
+        self.start_tempesta()
+        self.deproxy_manager.start()
+        self.assertTrue(deproxy_srv.wait_for_connections(timeout=1),
+                        "Cannot start Tempesta")
+
+    def tearDown(self):
+        tester.TempestaTest.tearDown(self)
+        self.sniffer.stop()
+        self.assertTrue(self.sniffer.check_results(), "Incorrect FIN-ACK sequence detected.")
+
+    def test_close_notify(self):
+        self.start_all()
+        tls_conn = TlsHandshake()
+        with tls_conn.socket_ctx():
+            self.assertTrue(tls_conn._do_12_hs(), "Can not connect to Tempesta")
+            res = tls_conn._do_12_req()
```

Comment:
The two lines above are exactly `do_12()` function... Why do we need to call the underlying functions directly?